keanu-project/src/main/java/io/improbable/keanu/algorithms/graphtraversal/DifferentiableChecker.java

```diff
@@ -0,0 +1,90 @@
+package io.improbable.keanu.algorithms.graphtraversal;
+
+import io.improbable.keanu.vertices.Vertex;
+import lombok.experimental.UtilityClass;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Set;
+
+@UtilityClass
+public class DifferentiableChecker {
+
+    /**
+     * Checks whether a set of vertices are differentiable
+     * w.r.t the latents in a graph.
+     *
+     * @param vertices the collection of vertices to check
+     * @return true if the vertices are differentiable w.r.t latents
+     */
+    public boolean isDifferentiable(Collection<Vertex> vertices) {
+        Queue<Vertex> queue = new LinkedList<>(vertices);
+        Set<Vertex> queued = new HashSet<>(vertices);
+        Set<Vertex> constantValueVerticesCache = new HashSet<>();
+
+        while (!queue.isEmpty()) {
+            Vertex visiting = queue.poll();
+
+            if (!visiting.isDifferentiable()) {
+                if (isVertexConstant(visiting, constantValueVerticesCache)) {
+                    continue;
+                } else {
+                    return false;
+                }
+            }
+
+            Collection<Vertex> nextVertices = visiting.getParents();
+            for (Vertex next : nextVertices) {
```

Comment:
If you're always passing in all latent and observed and trying to determine if the total logProb is differentiable then this will always work. If I wanted to know if the LogLikelihood function was differentiable then you'd have to not traverse past variables in this network and only consider the observed vertices. We should chat about this offline.

Reply:
Changed the algorithm based on our chat. Hopefully it should cover this case now.